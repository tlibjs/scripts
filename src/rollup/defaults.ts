export const inputBaseDir = "src";
export const outputBaseDir = "dist";
export const chunkFileNames = "_chunks/[name]-[hash].js";
