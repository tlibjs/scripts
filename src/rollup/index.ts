export * from "./bundle";
export * from "./dts";
export * from "./node";

export * from "./options";
