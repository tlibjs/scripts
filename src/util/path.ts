import { posix as pp } from "path";

export function joinPath(...paths: string[]): string {
  return pp.join(...paths);
}
