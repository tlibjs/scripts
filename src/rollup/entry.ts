import glob from "fast-glob";
import * as path from "path";

export interface EntryFileInfo {
  moduleName: string;
  fileRelativePath: string;
  index: number;
  baseDir: string;
}

export interface EntryFileFormatterOptions {
  /** whether to exclude baseDIr in returned filepaths */
  excludeBaseDir: boolean;
}

export type EntryFileFormatterSync = (
  info: EntryFileInfo,
  opts: EntryFileFormatterOptions,
) => readonly [string, string];

export type EntryFileFormatter = (
  ...args: Parameters<EntryFileFormatterSync>
) =>
  | Promise<ReturnType<EntryFileFormatterSync>>
  | ReturnType<EntryFileFormatterSync>;

export type MatchFilesPatterns = string | string[] | number;

export interface MatchFilesOptions {
  patterns: MatchFilesPatterns;
  baseDir: string;
  ignore: string | string[];
}

export interface GetEntryFilesOptionsCommon extends Partial<MatchFilesOptions> {
  keepIndexFiles?: boolean;
  excludeBaseDir?: boolean;
}

export interface GetEntryFilesOptions extends GetEntryFilesOptionsCommon {
  formatter?: EntryFileFormatter;
}

export interface GetEntryFilesOptionsSync extends GetEntryFilesOptionsCommon {
  formatter?: EntryFileFormatterSync;
}

function getGlobArgs({
  patterns,
  baseDir,
  ignore,
}: MatchFilesOptions): Parameters<typeof glob> {
  return [
    typeof patterns === "number" ? resolvePatternsByDepth(patterns) : patterns,
    {
      cwd: baseDir,
      ignore: typeof ignore === "string" ? [ignore] : ignore,
    },
  ];
}

export const DEFAULT_PATTERNS_EXT = "js,jsx,ts,tsx,json";
export const DEFAULT_PATTERNS = [
  `./*.{${DEFAULT_PATTERNS_EXT}}`,
  `./*/index.{${DEFAULT_PATTERNS_EXT}}`,
];

const DEFAULT_PATTERNS_WITHOUT_PREFIX = [
  `*.{${DEFAULT_PATTERNS_EXT}}`,
  `*/index.{${DEFAULT_PATTERNS_EXT}}`,
];

export function resolvePatternsByDepth(depth: number): string[] {
  if (depth === 1) return DEFAULT_PATTERNS;

  return new Array(depth)
    .fill(undefined)
    .map((_, i) =>
      DEFAULT_PATTERNS_WITHOUT_PREFIX.map((p) => `./${"*/".repeat(i)}${p}`),
    )
    .flat(1);
}

export const DEFAULT_IGNORE = [
  "**/*.d.ts",
  `**/*.{test,spec}.{${DEFAULT_PATTERNS_EXT}}`,
  `**/{__test__,__spec__}/**`,
];

function getModuleNameFromFile(info: path.ParsedPath) {
  return path.posix.join(info.dir, info.name);
}

function isSubDirIndexFile(info: path.ParsedPath) {
  const { name } = info;
  return name === "index" || name.startsWith("index.");
}

function getModuleNamesFromFiles(
  files: string[],
  keepIndexFiles: boolean,
): (readonly [moduleName: string, file: string])[] {
  files = files.map((f) => path.posix.normalize(f));

  const infoAndFiles = files.map((f) => [path.posix.parse(f), f] as const);

  if (keepIndexFiles) {
    return infoAndFiles.map(
      ([info, file]) => [getModuleNameFromFile(info), file] as const,
    );
  }

  const dirStatus = new Map<
    string,
    { hasIndex: boolean; hasOtherFiles: boolean }
  >();

  for (const [info] of infoAndFiles) {
    const { dir } = info;
    let status = dirStatus.get(dir);
    if (!status) {
      status = { hasIndex: false, hasOtherFiles: false };
      dirStatus.set(dir, status);
    }

    if (!status.hasIndex && isSubDirIndexFile(info)) {
      status.hasIndex = true;
    } else {
      status.hasOtherFiles = true;
    }
  }

  for (const subDir of dirStatus.keys()) {
    for (const [dir, status] of dirStatus) {
      if (subDir.startsWith(dir + "/")) {
        status.hasOtherFiles = true;
      }
    }
  }

  return infoAndFiles.map(([info, file]) => {
    const { dir } = info;
    const status = dirStatus.get(dir);
    if (dir !== "" && status && status.hasIndex && !status.hasOtherFiles) {
      // only process the dirs which only has one index file
      return [dir, file] as const;
    } else {
      return [getModuleNameFromFile(info), file] as const;
    }
  });
}

function validateFormatterReturnValue(
  newValue: readonly [string, string],
  warnPromise = false,
): readonly [string, string] {
  if (
    newValue &&
    typeof newValue[0] === "string" &&
    typeof newValue[1] === "string"
  ) {
    return newValue;
  } else {
    throw new Error(
      `getEntryFilesSync formatter is expected to return a [string, string], but returned ${
        warnPromise && newValue instanceof Promise
          ? "a Promise. Please consider using getEntryFiles"
          : String(newValue)
      }.`,
    );
  }
}

function tupleModuleNameFileToInfo(
  tuple: readonly [string, string],
  index: number,
  baseDir: string,
): EntryFileInfo {
  return {
    moduleName: tuple[0],
    fileRelativePath: tuple[1],
    index,
    baseDir,
  };
}

export function getEntryFilesSync({
  formatter,
  patterns = DEFAULT_PATTERNS,
  baseDir = "src",
  ignore = DEFAULT_IGNORE,
  keepIndexFiles = false,
  excludeBaseDir = false,
}: GetEntryFilesOptionsSync = {}): Record<string, string> {
  const files = glob.sync(
    ...getGlobArgs({
      patterns,
      baseDir,
      ignore,
    }),
  );

  let entries = getModuleNamesFromFiles(files, keepIndexFiles);

  if (formatter) {
    const opts = { excludeBaseDir };
    entries = entries
      .map((tuple, i) => tupleModuleNameFileToInfo(tuple, i, baseDir))
      .map((info) => {
        const newValue = formatter(info, opts);
        return validateFormatterReturnValue(newValue);
      });
  } else if (!excludeBaseDir && baseDir) {
    entries = entries.map(([mod, file]) => [
      mod,
      path.posix.join(baseDir, file),
    ]);
  }

  const entryFiles = Object.fromEntries(entries);

  return entryFiles;
}

export async function getEntryFiles({
  formatter,
  patterns = DEFAULT_PATTERNS,
  baseDir = "src",
  excludeBaseDir = false,
  ignore = DEFAULT_IGNORE,
  keepIndexFiles = false,
}: GetEntryFilesOptions = {}): Promise<Record<string, string>> {
  const files = await glob(
    ...getGlobArgs({
      patterns,
      baseDir,
      ignore,
    }),
  );

  let entries = getModuleNamesFromFiles(files, keepIndexFiles);

  if (formatter) {
    const opts = { excludeBaseDir };
    entries = await Promise.all(
      entries
        .map((tuple, i) => tupleModuleNameFileToInfo(tuple, i, baseDir))
        .map(async (info) => {
          const newValue = await formatter(info, opts);
          return validateFormatterReturnValue(newValue);
        }),
    );
  } else if (!excludeBaseDir && baseDir) {
    entries = entries.map(([mod, file]) => [
      mod,
      path.posix.join(baseDir, file),
    ]);
  }

  const entryFiles = Object.fromEntries(entries);

  return entryFiles;
}

export interface InferSingleEntryOptions {
  baseDir?: string;
}

export async function inferSingleEntry({
  baseDir,
}: InferSingleEntryOptions = {}): Promise<string> {
  const files = await glob(`./index.{${DEFAULT_PATTERNS_EXT}}`, {
    cwd: baseDir,
  });

  if (!files || files.length === 0) {
    throw new Error("failed to infer entry file");
  }

  if (files.length > 1) {
    throw new Error(
      `failed to infer entry file, multiple index.* files found: ${files.join(
        ",",
      )}`,
    );
  }

  const file = files[0];
  return baseDir ? path.posix.join(baseDir, file) : file;
}
