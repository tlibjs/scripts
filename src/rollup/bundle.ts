import { promises as fsp } from "fs";
import {
  commonPlugins,
  CommonPluginsOptions,
  compilePlugins,
} from "./common-plugins";
import { terser } from "rollup-plugin-terser";
import { pascalCase } from "pascal-case";
import commonjs from "@rollup/plugin-commonjs";
import { nodeResolve } from "@rollup/plugin-node-resolve";
import type { ModuleFormat, OutputOptions, RollupOptions } from "rollup";
import {
  resolve,
  resolveDict,
  Resolvable,
  ResolvableDict,
} from "../util/resolvable";
import { genOutputOptions, GenOutputOptions } from "./output";
import { inferSingleEntry } from "./entry";
import { joinPath } from "../util/path";
import * as DEFAULTS from "./defaults";

const COMMON_OUTPUT_OPTIONS: OutputOptions = {
  sourcemap: true,
  exports: "auto",
};

export const BUNDLE_FORMATS: ModuleFormat[] = [
  //
  "es",
  "iife",
  "umd",
];

export function genBundleOutput(
  format: ModuleFormat,
  { globalNamespace, min, outputRootDir }: RollupBundleOutputConfig,
): OutputOptions[] {
  // https://rollupjs.org/guide/en/#outputname
  const name = format === "es" ? undefined : globalNamespace;

  const minArr = min === "only" ? [true] : min ? [true, false] : [false];

  return minArr.map(
    (min): OutputOptions => ({
      format,
      file: joinPath(outputRootDir, `${format}${min ? ".min" : ""}.js`),
      name,
      plugins: min ? [terser()] : undefined,
    }),
  );
}

export async function inferGlobalNamespace(): Promise<string> {
  const json = await fsp.readFile("package.json", "utf-8");
  const pkg = JSON.parse(json) as { name?: unknown };
  const name = pkg.name;

  if (!name || typeof name !== "string") {
    throw new Error(
      "failed to inferGlobalNamespace: pkg.name is not a valid string",
    );
  }

  const namespace = pascalCase(name);
  return namespace;
}

export interface RollupBundleOutputConfig {
  min: boolean | "only";
  outputRootDir: string;
  /**
   * defauts to camel case of `pkg.name` (name field in CWD/package.json)
   */
  globalNamespace: string;
}

export interface RollupBundleOptions
  extends Partial<ResolvableDict<RollupBundleOutputConfig>>,
    CommonPluginsOptions {
  input?: Resolvable<string | undefined>;
  output?: Resolvable<GenOutputOptions, [RollupBundleOutputConfig]>;
  inputBaseDir?: string;
  outputBaseDir?: string;
}

export function genBundleOutputOptions(
  conf: RollupBundleOutputConfig,
): RollupOptions[] {
  return BUNDLE_FORMATS.map((format) => genBundleOutput(format, conf)).flat(1);
}

export async function rollupBundle({
  inputBaseDir = DEFAULTS.inputBaseDir,
  outputBaseDir = DEFAULTS.outputBaseDir,
  input,
  output: _output = genBundleOutputOptions,
  outputRootDir = "bundle",
  min = false,
  globalNamespace = inferGlobalNamespace,
  ...commonPluginsOpts
}: RollupBundleOptions = {}): Promise<RollupOptions> {
  const { input: inputFile, ...conf } = await resolveDict({
    input,
    outputRootDir,
    min,
    globalNamespace,
  });

  const genOutput = await resolve(_output, {
    ...conf,
    outputRootDir: joinPath(outputBaseDir, conf.outputRootDir),
  });

  const output = genOutputOptions(genOutput, COMMON_OUTPUT_OPTIONS);

  return {
    input: inputFile ?? (await inferSingleEntry({ baseDir: inputBaseDir })),
    output,
    plugins: [
      //
      ...compilePlugins(),
      nodeResolve(),
      commonjs(),
      ...commonPlugins({ outputBaseDir, ...commonPluginsOpts }),
    ],
  };
}
