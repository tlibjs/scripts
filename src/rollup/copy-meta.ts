import { PluginImpl } from "rollup";
import glob from "fast-glob";
import * as fs from "fs";
import { promises as fsp } from "fs";
import * as path from "path";

export const COPY_META_DEFAULT_SOURCE = ["README*", "LICENSE*", "CHANGELOG*"];

export interface CopyMetaPluginOptions {
  to: string;
  source?: string | string[];
  caseSensitiveMatch?: boolean;
  overwriteExisting?: boolean;
}

export const copyMeta: PluginImpl<CopyMetaPluginOptions> = (
  {
    //
    to,
    source = COPY_META_DEFAULT_SOURCE,
    caseSensitiveMatch = false,
    overwriteExisting = false,
  } = { to: "" },
) => ({
  name: "@tlib/rollup-plugin-copy-meta",
  buildEnd: async () => {
    if (!to) throw new Error("copyMeta options.dir cannot be falsy");

    await fsp.mkdir(to, { recursive: true });

    const files = await glob(source, {
      onlyFiles: true,
      caseSensitiveMatch,
      unique: true,
    });

    const mode = overwriteExisting ? undefined : fs.constants.COPYFILE_EXCL;

    await Promise.all(
      files.map((file) => {
        const filename = path.basename(file);
        const dest = path.join(to, filename);
        return fsp.copyFile(file, dest, mode);
      }),
    );
  },
});
