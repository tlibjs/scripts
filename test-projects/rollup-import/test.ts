import * as path from "path";
import { posix as pp } from "path";
import { rollup } from "rollup";
import nodeResolve from "@rollup/plugin-node-resolve";

const FORMATS = ["esm", "cjs"] as const;

test("rollup build", async () => {
  const bundle = await rollup({
    input: path.join(__dirname, "./index.mjs"),
    plugins: [nodeResolve()],
  });

  expect(
    bundle.watchFiles.map((f) => pp.relative(__dirname, f)),
  ).toMatchSnapshot("watchFiles");

  const outputs = await Promise.all(
    FORMATS.map((format) =>
      bundle.generate({ format }).then((out) => {
        const outs = out.output.map((output) => {
          return {
            type: output.type,
            name: output.name,
            fileName: output.fileName,
            code: output.type === "chunk" ? output.code : null,
          };
        });

        return [format, outs];
      }),
    ),
  );

  expect(Object.fromEntries(outputs)).toMatchSnapshot("outputs");
});
