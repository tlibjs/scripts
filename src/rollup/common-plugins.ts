import typescript, { RollupTypescriptOptions } from "@rollup/plugin-typescript";
import babel, { RollupBabelInputPluginOptions } from "@rollup/plugin-babel";
import json from "@rollup/plugin-json";
import type { Plugin } from "rollup";
import { copyMeta, CopyMetaPluginOptions } from "./copy-meta";
import { isTruthy } from "../util/func";
import * as DEFAULTS from "./defaults";

export interface CompilePluginsOptions {
  typescript?: Partial<RollupTypescriptOptions>;
  babel?: Partial<RollupBabelInputPluginOptions>;
}

export function compilePlugins(options: CompilePluginsOptions = {}): Plugin[] {
  return [
    // https://www.npmjs.com/package/@rollup/plugin-json
    json(),
    // https://github.com/rollup/plugins/tree/master/packages/typescript
    typescript({
      tsconfig: "tsconfig.json",
      sourceMap: true,
      inlineSources: true,
      include: "src/**/*.ts",
      exclude: ["**/*.spec.ts", "**/*.test.ts", "**/__test__/**/*.spec.ts"],
      ...options.typescript,
    }),
    // https://github.com/rollup/plugins/tree/master/packages/babel
    babel({
      extensions: [".js", ".jsx", ".es6", ".es", ".mjs", ".ts", ".tsx"],
      babelHelpers: "bundled",
      exclude: /node_modules/,
      ...options.babel,
    }),
  ];
}

export interface CommonPluginsOptions {
  outputBaseDir?: string;
  /**
   * whether to copy meta files to outputBaseDir.
   *
   * e.g. `README*` `CHANGELOG*` `LICENSE*`
   */
  copyMeta?: boolean | CopyMetaPluginOptions;
}

export function commonPlugins({
  outputBaseDir = DEFAULTS.outputBaseDir,
  copyMeta: copyMetaOpts,
}: CommonPluginsOptions = {}): Plugin[] {
  return [
    copyMetaOpts
      ? copyMeta(copyMetaOpts === true ? { to: outputBaseDir } : copyMetaOpts)
      : null,
  ].filter(isTruthy);
}
