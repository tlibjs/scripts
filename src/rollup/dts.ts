import type { InputOption, RollupOptions } from "rollup";
import * as DEFAULTS from "./defaults";
import { getEntryFiles } from "./entry";
import { Resolvable, resolve } from "../util/resolvable";
import { commonPlugins, CommonPluginsOptions } from "./common-plugins";
import json from "@rollup/plugin-json";

export interface RollupDtsOptions extends CommonPluginsOptions {
  inputBaseDir?: string;
  outputBaseDir?: string;
  input?: Resolvable<InputOption | undefined>;
}

export async function rollupDts({
  inputBaseDir = DEFAULTS.inputBaseDir,
  outputBaseDir = DEFAULTS.outputBaseDir,
  input,
  ...commonPluginsOpts
}: RollupDtsOptions = {}): Promise<RollupOptions> {
  const dts = (await import("rollup-plugin-dts")).default;

  const inputOption =
    (await resolve(input)) ?? (await getEntryFiles({ baseDir: inputBaseDir }));

  return {
    input: inputOption,
    output: {
      dir: outputBaseDir,
      format: "es",
      // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
      chunkFileNames: (info) => {
        const name = info.name.replace(/\.d$/, "");
        return DEFAULTS.chunkFileNames
          .replace(/\[name\]/g, name)
          .replace(/\.js$/, ".d.ts");
      },
    },
    plugins: [
      //
      json(),
      dts(),
      ...commonPlugins({ outputBaseDir, ...commonPluginsOpts }),
    ],
  };
}
