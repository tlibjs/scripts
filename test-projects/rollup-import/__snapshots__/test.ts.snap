// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rollup build: outputs 1`] = `
Object {
  "cjs": Array [
    Object {
      "code": "'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var path = require('path');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n[\\"default\\"] = e;
  return Object.freeze(n);
}

var path__namespace = /*#__PURE__*/_interopNamespace(path);

/**
 * This module is only for testing
 */
const MESSAGE = \\"This module is only for testing\\";
function hello(name = \\"world\\") {
  return \\"Hello \\".concat(name, \\"!\\");
}
function testPathDelimiter() {
  return path__namespace.delimiter;
}

var testMod = /*#__PURE__*/Object.freeze({
  __proto__: null,
  MESSAGE: MESSAGE,
  hello: hello,
  testPathDelimiter: testPathDelimiter
});

exports.testMod = testMod;
",
      "fileName": "index.js",
      "name": "index",
      "type": "chunk",
    },
  ],
  "esm": Array [
    Object {
      "code": "import * as path from 'path';

/**
 * This module is only for testing
 */
const MESSAGE = \\"This module is only for testing\\";
function hello(name = \\"world\\") {
  return \\"Hello \\".concat(name, \\"!\\");
}
function testPathDelimiter() {
  return path.delimiter;
}

var testMod = /*#__PURE__*/Object.freeze({
  __proto__: null,
  MESSAGE: MESSAGE,
  hello: hello,
  testPathDelimiter: testPathDelimiter
});

export { testMod };
",
      "fileName": "index.js",
      "name": "index",
      "type": "chunk",
    },
  ],
}
`;

exports[`rollup build: watchFiles 1`] = `
Array [
  "index.mjs",
  "../../dist/es/util/test-mod.js",
]
`;
